### PR TITLE
fix: reproduce VisDrone and SKU-110K with v0.1-N and EsMoE-N

### PR DESCRIPTION
**Pull request by miangder**
# Rhinoceros-Bird Task: VisDrone & SKU-110K Reproduction Results

This reproduction uses the `YOLO-Master-v0.1-N.pt` and `YOLO-Master-EsMoE-N.pt` checkpoints released with YOLO-Master-v26.02, together with the built-in dataset configurations:

- `ultralytics/cfg/datasets/VisDrone.yaml`
- `ultralytics/cfg/datasets/SKU-110K.yaml`

All four training runs completed 120 epochs. Epochs 1--103 of SKU-110K EsMoE-N were completed on the old node. A GPU 2 driver failure interrupted training, so epochs 104--120 were resumed on an RTX 5090 node. The two original `results.csv` segments were merged into a complete 120-epoch record, and the epoch-120 `last.pt` was validated independently. Every run preserves per-epoch `mAP50`, `mAP50-95`, `box_loss`, `cls_loss`, and `moe_loss`. This report contains only observed log results.

## Added Files

- `scripts/reproduce/reproduce_visdrone.py`: supports `--model v01` and `--model moe`.
- `scripts/reproduce/reproduce_sku110k.py`: supports `--model v01` and `--model moe`.
- `scripts/reproduce/common.py`: shared training, offline W&B, per-epoch aggregation, and Windows `workers=0` compatibility logic.
- `README_reproduce.md`: this report.

## Environment

- Original remote hardware: 6 x NVIDIA GeForce RTX 2080 Ti, 11 GB VRAM per GPU; GPUs 0, 1, and 2 were used.
- NVIDIA Driver: 580.126.09.
- Python: 3.8.10.
- PyTorch: 2.1.2+cu121.
- Ultralytics: 8.3.240.
- Training settings: `batch=4`, `workers=4`, `patience=0`, W&B offline mode.

The resume stage for SKU-110K EsMoE-N ran on a new node with an NVIDIA GeForce RTX 5090 (32 GB), Driver 580.76.05, Python 3.12.3, and PyTorch 2.11.0+cu128. Its training parameters remained `imgsz=640`, `batch=4`, `workers=4`, and `patience=0`.

The official release checkpoints contain 7.555M parameters for YOLO-Master-v0.1-N and 2.694M parameters for YOLO-Master-EsMoE-N. The number of dataset classes changes the detection head, so the parameter count shown by the training-time model summary can differ slightly.

## Dataset Download

Run the following from the repository root. The YAML files contain dataset download and annotation-conversion logic:

```bash
python scripts/reproduce/reproduce_visdrone.py --download-only --wandb disabled
python scripts/reproduce/reproduce_sku110k.py --download-only --wandb disabled
```

When the dataset root is not in the default location, configure it first:

```bash
yolo settings datasets_dir=/path/to/datasets
```

## Training Commands

The following are the equivalent commands used for this reproduction. The scripts obtain the corresponding official YOLO-Master-v26.02 release checkpoints by default. Offline servers can point `--v01-model` or `--esmoe-model` at an already downloaded `.pt` file.

```bash
# VisDrone baseline, imgsz=800
python scripts/reproduce/reproduce_visdrone.py --model v01 --epochs 120 --imgsz 800 --batch 4 --device 0 --workers 4 --patience 0 --wandb offline

# VisDrone EsMoE-N, imgsz=800
python scripts/reproduce/reproduce_visdrone.py --model moe --epochs 120 --imgsz 800 --batch 4 --device 1 --workers 4 --patience 0 --wandb offline

# SKU-110K baseline, imgsz=640
python scripts/reproduce/reproduce_sku110k.py --model v01 --epochs 120 --imgsz 640 --batch 4 --device 0 --workers 4 --patience 0 --wandb offline

# SKU-110K EsMoE-N, imgsz=640
python scripts/reproduce/reproduce_sku110k.py --model moe --epochs 120 --imgsz 640 --batch 4 --device 2 --workers 4 --patience 0 --wandb offline
```

## Reproduction Results

The near-zero mAP from the original EsMoE-N run was traced to an inference-path error: validation still pruned experts when `top_k` equaled the expert count. After the fix, VisDrone EsMoE-N completed all 120 epochs and its validation metrics returned to normal. The table below uses the final row of each `results.csv` file.

| Dataset | Model | Input Resolution | Epochs | mAP50 | mAP50-95 | Release Parameters | Status |
| --- | --- | ---: | ---: | ---: | ---: | ---: | --- |
| VisDrone | YOLO-Master v0.1-N | 800 | 120 | 0.42966 | 0.26027 | 7.555M | Valid |
| VisDrone | YOLO-Master EsMoE-N | 800 | 120 | 0.42482 | 0.25704 | 2.694M | Valid |
| SKU-110K | YOLO-Master v0.1-N | 640 | 120 | 0.91243 | 0.59063 | 7.555M | Valid |
| SKU-110K | YOLO-Master EsMoE-N | 640 | 120 | 0.91344 | 0.59013 | 2.694M | Valid |

## Training Logs

Complete training logs, per-epoch CSV files, training arguments, state files, plots, and offline W&B run packages can be downloaded from Baidu Netdisk:

- Shared folder: Offline Logs
- Link: https://pan.baidu.com/s/16IsBfEuPtp7hwnhE-rzOUw?pwd=n5vy
- Extraction code: `n5vy`

## Known Issues and Solutions

- When `top_k == num_experts`, EsMoE-N incorrectly entered a thresholded sparse-pruning path, causing near-zero validation mAP. `ES_MOE._sparse_forward()` was changed to use dense computation consistent with training; the VisDrone rerun completed and validated normally.
- SKU-110K EsMoE-N triggered `CUDA unspecified launch failure` on GPU 2 at epoch 103. PyTorch could not initialize any CUDA device on the old node afterward. The resume point was moved to an RTX 5090 (32 GB, Driver 580.76.05, PyTorch 2.11.0+cu128); the dataset was fully verified with `rsync --partial --append-verify`, and the remaining 17 epochs completed successfully with `--resume --device 0`.
- The original validator lacked `LOCAL_RANK` and `torch_distributed_zero_first` imports and called the nonexistent `convert_ndjson_to_yolo_if_needed`. It was fixed in `ultralytics/engine/validator.py` to invoke the available converter only for `.ndjson`. All four `best.pt` checkpoints were independently validated successfully.
- Some SKU-110K JPEG files emit truncated or corrupted-image warnings. Ultralytics repairs or skips unreadable samples; the warnings are retained in the logs.
- Two training instances can race while first creating the same `labels.cache` file. Build the cache once first, or start the first epoch sequentially before parallel runs, to avoid `UnpicklingError`.
- On Windows, use `--workers 0`; multiprocessing data loading can otherwise cause an I/O deadlock.
- When VRAM is insufficient, lower `--batch` first. This reproduction used `batch=4` with 11 GB VRAM.

### CODE REVIEW

**Comment by isLinXu** [comment]
LGTM
